plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ScanQueryPageSource.java

```diff
@@ -83,10 +86,18 @@ public ScanQueryPageSource(
         // Columns to fetch as doc_fields instead of pulling them out of the JSON source
         // This is convenient for types such as DATE, TIMESTAMP, etc, which have multiple possible
         // representations in JSON, but a single normalized representation as doc_field.
-        List<String> documentFields = flattenFields(columns).entrySet().stream()
-                .filter(entry -> entry.getValue().equals(TIMESTAMP_MILLIS))
-                .map(Map.Entry::getKey)
-                .collect(toImmutableList());
+        List<String> documentFields = new ArrayList<>();
+        Set<String> docFieldsToRemove = new HashSet<>();
+        columns.stream()
+                .filter(ch -> ch.isSupportsPredicates() || ch.isPredicateWithSubFields())
+                // array and json does not support documnet fields,
+                // The order of the two results is inconsistent between the source field and the docvalue field
+                // see https://discuss.elastic.co/t/query-array-of-keyword-responce-different-in-source-and-docfields/302389
+                .filter(ch -> !ch.isArray() && !ch.asRawJson())
+                .forEach(ch -> {
+                    documentFields.add(ch.getPredicateName());
```

Comment:
do we need to use the predicateName here for projection? It seems like we're implicitly saying that multifield has the same decoder, which may not be true.